numpy/core/src/multiarray/nditer_constr.c

```diff
@@ -2920,10 +2927,11 @@ npyiter_allocate_arrays(NpyIter *iter,
                     return 0;
                 }
             }
-            /* If the data will be written to, set UPDATEIFCOPY */
+            /* If the data will be written to, set WRITEBACKIFCOPY
+               and require a context manager */
             if (op_itflags[iop] & NPY_OP_ITFLAG_WRITE) {
                 Py_INCREF(op[iop]);
-                if (PyArray_SetUpdateIfCopyBase(temp, op[iop]) < 0) {
+                if (PyArray_SetWritebackIfCopyBase(temp, op[iop]) < 0) {
                     Py_DECREF(temp);
                     return 0;
                 }
```

Comment:
In this file, in `NpyIter_AdvancedNew`, we need some calls to `PyArray_DiscardWritebackIfCopy` if a failure occurs after the call to `npyiter_allocate_arrays`.

Reply:
Pinging this one again. I still think we need some calls to `PyArray_DiscardWritebackIfCopy`(2, I think) in `NpyIter_AdvancedNew`, if a failure occurs after the call to `npyiter_allocate_arrays`.

Reply:
fixed, added ``NpyIter_Close`